File: numbers/happy_num/ruby/happy_num.rb
#!/usr/bin/ruby
#
# Created:  Mon 05 Jun 2017 10:55:34 PM PDT
# Modified: Thu 08 Jun 2017 12:41:01 PM PDT
#
# Copyright 2017 (C) Robert Gill
#

class Integer
  def happy?
    num = self
    loop do
      num = sum_sqr_digits(num)
      break if num == 1 || num < 10
    end
    num == 1
  end

  private

  def sum_sqr_digits(num)
    sum = 0
    while num > 0
      digit = num % 10
      sum += digit * digit
      num /= 10
    end
    sum
  end
end

found = 0
2.step.each do |n|
  if n.happy?
    found += 1
    puts "#{n} is a Happy Number!"
  end
  break if found == 8
end

File: numbers/happy_num/c/happy_num.c
/*
 * Created:  Tue 30 May 2017 12:43:28 PM PDT
 * Modified: Sun 11 Jun 2017 06:20:58 PM PDT
 *
 * Copyright 2017 (C) Robert Gill
 *
 */

#include <stdio.h>
#include <stdlib.h>

int
sum_sqr_digits (int num)
{
  int digit;
  int sum = 0;

  while (num > 0)
    {
      digit = num % 10;
      sum += digit * digit;
      num /= 10;
    }

  return sum;
}

int
is_happy (int num)
{
  do num = sum_sqr_digits (num);
  while (num != 1 && num > 9);
  return (num == 1);
}

int
main (int argc, char **argv)
{
  int count = 0, num = 2;

  while (count < 8)
    {
      if (is_happy (num))
        {
          printf ("%d is a Happy Number!\n", num);
          count += 1;
        }
      num += 1;
    }

  return EXIT_SUCCESS;
}
